sim/core/character.go

```diff
@@ -229,21 +241,61 @@ func (character *Character) Advance(sim *Simulation, elapsedTime time.Duration)
 
 // Returns rate of mana regen, as mana / second
 func (character *Character) manaRegenPerSecond() float64 {
-	return character.stats[stats.MP5] / 5.0
+	regen := character.stats[stats.MP5] / 5.0
+	if character.PsuedoStats.SpiritRegenRateCasting > 0 {
+		// Since spirit formula is expensive I just added this if here.
+		//  TODO: We could also add some stat caches on calculated things like regen and spell hit etc and not need this check
+		bonus := (character.PsuedoStats.SpiritRegenRateCasting * character.spiritManaRegen()) * character.PsuedoStats.ManaRegenMultiplier
+		regen += bonus
+	}
+	return regen
+}
+
+func (character *Character) spiritManaRegen() float64 {
+	return 0.001 + character.stats[stats.Spirit]*math.Sqrt(character.stats[stats.Intellect])*0.009327
 }
 
 // Regenerates mana based on MP5 stat and the elapsed time. This function
 // ignores Spirit-based regen; only use it to improve performance in cases where
-// there is a gaurantee to have no spirit regen.
+// there is a gaurantee to have no spirit regen (outside of 5s rule).
 func (character *Character) RegenManaMP5Only(sim *Simulation, elapsedTime time.Duration) {
 	// MP5 regen
 	regen := character.manaRegenPerSecond() * elapsedTime.Seconds()
-	character.stats[stats.Mana] += regen
+	character.addMana(regen)
+	if sim.Log != nil && regen != 0 {
+		sim.Log(" Advanced [%0.1fs] Regenerated: %0.1f mana. Total: %0.1f\n", elapsedTime.Seconds(), regen, character.CurrentMana())
+	}
+}
+
+func (character *Character) addMana(amount float64) {
+	character.stats[stats.Mana] += amount
 	if character.CurrentMana() > character.MaxMana() {
 		character.stats[stats.Mana] = character.MaxMana()
 	}
+}
+
+// Regenerates mana using mp5 and spirit. Will calculate time since last cast and then enable spirit regen if needed.
+func (character *Character) FullRegen(sim *Simulation, elapsedTime time.Duration) {
+	var regen float64
+	if character.PsuedoStats.FiveSecondRuleRefreshTime >= sim.CurrentTime {
+		// this means we fully in spirit regen.
+		regen = character.stats[stats.MP5]/5.0 + character.spiritManaRegen()*elapsedTime.Seconds()
+	} else if sim.CurrentTime+elapsedTime > character.PsuedoStats.FiveSecondRuleRefreshTime {
```

Comment:
I think at some point I might have changed the assignment to sim.CurrentTime to happen at the beginning of Advance() so this should just be sim.CurrentTime, and the previous check should be sim.CurrentTime-elapsedTime. Or maybe it would be best to be super explicit, and change all the Advance() functions to take (sim, oldTime, elapsedTime, newTime)

Reply:
you are correct, i fixed all the code i think